firedrake/slate/slate.py

```diff
@@ -0,0 +1,767 @@
+        return self.__sub__(other)
+
+    def __mul__(self, other):
+        return TensorMul(self, other)
+
+    def __rmul__(self, other):
+        """Tensor multiplication is not commutative in general."""
+        return TensorMul(other, self)
+
+    def __neg__(self):
+        return Negative(self)
+
+    def __pos__(self):
+        return Positive(self)
+
+    def equals(self, other):
+        """Evaluates the boolean expression `A == B`."""
+
+        if not isinstance(other, Tensor):
+            return False
+        if len(self.arguments()) != len(other.arguments()):
+            return False
+        if len(self.coefficients()) != len(other.coefficients()):
+            return False
+        if len(self._integrals) != len(other._integrals):
+            return False
+        return (self._arguments == other._arguments and
+                self._coefficients == other._coefficients and
+                self._integrals == other._integrals)
+
+    def __ne__(self, other):
+        return not self.equals(other)
+
+    def __eq__(self, other):
+        """Evaluation of the "==" operator using the SLATE class: SlateEquation."""
+        return SlateEquation(self, other)
+
+    # Essential properties
+    @property
+    def inv(self):
+        return Inverse(self)
+
+    @property
+    def T(self):
+        return Transpose(self)
+
+    @property
+    def operands(self):
+        """Returns the tensor objects which this tensor operates on."""
+        return ()
+
+    # Analyze integrals of tensor
+    @classmethod
+    def check_integrals(cls, integrals):
+        """Checks for positive restrictions on integrals."""
+
+        mapper = CheckRestrictions()
+        for it in integrals:
+            map_integrand_dags(mapper, it)
+
+    # Accessor methods
+    def arguments(self):
+        """Returns a tuple of arguments associated with the tensor."""
+        return self._arguments
+
+    def coefficients(self):
+        """Returns a tuple of coefficients associated with the tensor."""
+        return self._coefficients
+
+    def get_ufl_integrals(self):
+        """Returns the associated ufl integrals with the tensor
+        (This is necessary for form compilation).
+
+        This function should really only be called on base-Tensor objects
+        like an individual Vector or Matrix rather than a full SLATE expr."""
+        return self._integrals
+
+    def integrals(self):
+        """Returns a tuple of SlateIntegrals associated with the tensor."""
+        return self._slate_integrals
+
+    def integrals_by_type(self, integral_type):
+        """Returns a tuple of integrals corresponding with a particular domain type."""
+
+        slate_assert(integral_type in ['cell', 'interior_facet', 'exterior_facet'],
+                     "Integral type %s is not supported." % integral_type)
+        return tuple(it for it in self.integrals() if it.integral_type() == integral_type)
+
+    def ufl_domains(self):
```

Comment:
OK, so now you're going to go ahead and implement a load of methods that `ufl.Form` has, but you're not inheriting because these guys aren't actually fully-fledged forms.